frontend/readhistory.lua

```diff
@@ -227,63 +212,64 @@ function ReadHistory:fileSettingsPurged(path)
 end
 
 function ReadHistory:clearMissing()
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == nil or lfs.attributes(self.hist[i].file, "mode") ~= "file" then
-            self:removeItem(self.hist[i], i)
+    local history_updated
+    for i, v in ipairs(self.hist) do
+        if v.file == nil or lfs.attributes(v.file, "mode") ~= "file" then
+            self:removeItem(v, i, true) -- no flush
+            history_updated = true
         end
     end
-    self:ensureLastFile()
+    if history_updated then
+        self:_flush()
+        self:ensureLastFile()
+    end
 end
 
 function ReadHistory:removeItemByPath(path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == path then
-            self:removeItem(self.hist[i])
-            break
-        end
+    local index = self:getIndexByFile(path)
+    if index then
+        self:removeItem(self.hist[index], index)
     end
-    self:ensureLastFile()
 end
 
 function ReadHistory:updateItemByPath(old_path, new_path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == old_path then
-            self.hist[i].file = new_path
-            self.hist[i].text = new_path:gsub(".*/", "")
-            self:_flush()
-            self:reload(true)
-            self.hist[i].callback = function()
-                selectCallback(new_path)
-            end
-            break
-        end
+    local index = self:getIndexByFile(old_path)
+    self.hist[index].file = new_path
+    self.hist[index].text = new_path:gsub(".*/", "")
+    self.hist[index].callback = function()
+        selectCallback(new_path)
     end
+    self:_flush()
+    self:reload(true)
     if G_reader_settings:readSetting("lastfile") == old_path then
         G_reader_settings:saveSetting("lastfile", new_path)
     end
     self:ensureLastFile()
 end
 
-function ReadHistory:removeItem(item, idx)
-    table.remove(self.hist, item.index or idx)
+function ReadHistory:removeItem(item, idx, no_flush)
+    local index = idx or self:getIndexByTime(item.time)
+    table.remove(self.hist, index)
     os.remove(DocSettings:getHistoryPath(item.file))
-    self:_indexing(item.index or idx)
-    self:_flush()
-    self:ensureLastFile()
+    if not no_flush then
```

Comment:
I remember we had a talk some months ago about too many flush of something when you would act on many files, possibly with your added multi selection.
Does this no_fluxh is about fixing it ? (I see it's use in clearMissing() only.)

Reply:
It is applpied when removing many (deleted) items from the History list only.

In FM it is not applicable, currently we manipulate on each of the selected files one by one.